### PR TITLE
Add Portfolio Optimisation Module

### PR DESCRIPTION
**Pull request by aditya1702**
# Description

Adds a portfolio optimisation module in the package. Currently this module houses the following algorithms - 
1. CLA
2. HRP
3. Inverse Variance Portfolio

## Type of change
Add main code files for the algorithms and their respective unittests

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Wrote unit tests for all code

**Test Configuration**:
* Operating system: Mac OSX
* IDE used: Pycharm

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Jackal08** [comment]
The following changes need to please be made:
* When ready make PR to dev, not master.
* Complete Unit tests for 100% code coverage
* Make sure linter passes all inspection
* Make sure to update your branch with dev before creating PR

I am closing this PR since it needs to point to dev.

**Comment by imcu** [comment]
I closed this PR as it needs to point to `dev` branch and not `master`

**Comment by aditya1702** [comment]
> I closed this PR as it needs to point to `dev` branch and not `master`

I am already pointing it to dev

File: .gitignore [hunk 1]
```diff
@@ -1,9 +1,12 @@
 .idea/*
+*.pyc
+*.txt
```
Comment (Jackal08) [inline .gitignore]: No please dont ignore .txt files. The requirements file is a txt file.

File: mlfinlab/portfolio_optimization/hrp.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+import numpy as np
+import pandas as pd
+from scipy.cluster.hierarchy import dendrogram, linkage
+from scipy.spatial.distance import squareform
+import matplotlib.pyplot as plt
+from sklearn.covariance import OAS
+
+
+class HierarchicalRiskParity:
+
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/hrp.py]: Missing Class docstring

File: mlfinlab/portfolio_optimization/cla.py [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+import pandas as pd
+from math import log, ceil
+
+
+def _infnone(x):
+    return float("-inf") if x is None else x
+
+
+class CLA:
+    def __init__(self, weight_bounds=(0, 1), calculate_returns="mean"):
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/cla.py]: Missing class docstring

File: mlfinlab/portfolio_optimization/cla.py [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+'''
+
+import numbers
+import numpy as np
+import pandas as pd
+from math import log, ceil
+
+
+def _infnone(x):
+    return float("-inf") if x is None else x
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/cla.py]: Docstring

File: mlfinlab/portfolio_optimization/cla.py [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+        '''
+
+        return matrix[np.ix_(listX, listY)]
+
+    def _purge_num_err(self, tol):
+        '''
+        Purge violations of inequality constraints (associated with ill-conditioned cov matrix)
+        '''
+
+        i = 0
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/cla.py]: Inline comments please.

File: mlfinlab/portfolio_optimization/cla.py [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+                del self.free_weights[i]
+            else:
+                i += 1
+
+    def _purge_excess(self):
+        '''
+        Remove violations of the convex hull
+        '''
+
+        i, repeat = 0, False
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/cla.py]: inline comments

File: mlfinlab/portfolio_optimization/cla.py [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+        '''
+
+        w = a * w0 + (1 - a) * w1
+        b = np.dot(w.T, self.expected_returns)[0, 0]
+        c = np.dot(np.dot(w.T, self.cov_matrix), w)[0, 0] ** 0.5
+        return b / c
+
+    def _bound_free_weight(self, free_weights):
+        '''
+        Add a free weight to list of bounded weights
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/cla.py]: missing params in docstring

File: mlfinlab/portfolio_optimization/cla.py [hunk 8]
```diff
@@ -0,0 +1,538 @@
...
+                l, bi = self._compute_lambda(
+                    covarF_inv, covar_FB, mean_F, w_B, j, [self.lower_bounds[i], self.upper_bounds[i]]
+                )
+                if _infnone(l) > _infnone(lambda_in):
+                    lambda_in, i_in, bi_in = l, i, bi
+                j += 1
+        return lambda_in, i_in, bi_in
+
+    def _free_bound_weight(self, free_weights):
+        '''
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/cla.py]: missing params in docstring

File: mlfinlab/portfolio_optimization/cla.py [hunk 9]
```diff
@@ -0,0 +1,538 @@
...
+                    w_B,
+                    mean_F.shape[0] - 1,
+                    self.weights[-1][i],
+                )
+                if (self.lambdas[-1] is None or l < self.lambdas[-1]) and l > _infnone(lambda_out):
+                    lambda_out, i_out = l, i
+        return lambda_out, i_out
+
+    def _initialise(self, asset_prices):
+
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/cla.py]: Docstring

File: mlfinlab/portfolio_optimization/cla.py [hunk 10]
```diff
@@ -0,0 +1,538 @@
...
+        '''
+        Compute the entire efficient frontier solution
+
+        :param points: (int) number of efficient frontier points to be calculated
+        :return: tuple of mean, variance amd weights of the frontier solutions
+        '''
+
+        mu, sigma, weights = [], [], []
+
+        # remove the 1, to avoid duplications
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/cla.py]: Capital letters

File: mlfinlab/portfolio_optimization/hrp.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+    def __init__(self):
+        return
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/hrp.py]: Empty return? Do we need an init func then?

File: mlfinlab/portfolio_optimization/hrp.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+        :param height: (int) height of the plot
+        :param width: (int) width of the plot
+        '''
+        plt.figure(figsize=(width, height))
+        dendrogram(self.clusters)
+        plt.show()
+
+    def _calculate_returns(self, asset_prices, resample_returns_by):
+        '''
+
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/hrp.py]: Docstring

File: mlfinlab/portfolio_optimization/hrp.py [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+        asset_returns = asset_prices.pct_change()
+        asset_returns = asset_returns.dropna(how='all')
+        asset_returns = asset_returns.resample(resample_returns_by).mean()
+        return asset_returns
+
+    def _shrink_covariance(self, covariance):
+        '''
+
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/hrp.py]: Docstring

File: mlfinlab/portfolio_optimization/hrp.py [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+        :return: (pd.Dataframe) shrinked asset returns covariances
+        '''
+
+        oas = OAS()
+        oas.fit(covariance)
+        shrinked_covariance = oas.covariance_
+        return pd.DataFrame(shrinked_covariance, index=covariance.columns, columns=covariance.columns)
+
+    def _cov2corr(self, covariance):
+        '''
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/hrp.py]: Docsting

File: mlfinlab/portfolio_optimization/mean_variance.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+This module implements the classic mean-variance optimisation techniques for calculating the efficient frontier.
+It uses typical quadratic optimisers to generate optimal portfolios for different objective functions.
+'''
+
+import numpy as np
+import pandas as pd
+
+
+class MeanVarianceOptimisation:
+
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/mean_variance.py]: Class docstring

File: mlfinlab/portfolio_optimization/mean_variance.py [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+    def __init__(self):
+        return
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/mean_variance.py]: Do we need this?

File: mlfinlab/portfolio_optimization/mean_variance.py [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+
+    def allocate(self, asset_prices, solution='inverse_variance', resample_returns_by='B'):
+        '''
+
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/mean_variance.py]: docstring

File: mlfinlab/portfolio_optimization/mean_variance.py [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+        self.weights = []
+        if solution == 'inverse_variance':
+            self.weights = self._inverse_variance(asset_returns=asset_returns)
+        self.weights = pd.DataFrame(self.weights)
+        self.weights.index = assets
+        self.weights = self.weights.T
+
+    def _calculate_returns(self, asset_prices, resample_returns_by):
+        '''
+
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/mean_variance.py]: docstring

File: mlfinlab/portfolio_optimization/mean_variance.py [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+        asset_returns = asset_prices.pct_change()
+        asset_returns = asset_returns.dropna(how='all')
+        asset_returns = asset_returns.resample(resample_returns_by).mean()
+        return asset_returns
+
+    def _inverse_variance(self, asset_returns):
+        '''
+
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/mean_variance.py]: docstring

File: mlfinlab/portfolio_optimization/mean_variance.py [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+        :return: (np.array) array of portfolio weights
+        '''
+
+        cov = asset_returns.cov()
+        ivp = 1. / np.diag(cov)
+        ivp /= ivp.sum()
+        return ivp
+
+    def _min_volatility(self):
+        return
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/mean_variance.py]: Do we need this?

File: mlfinlab/portfolio_optimization/mean_variance.py [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+
+    def _max_sharpe(self):
+        return
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/mean_variance.py]: Do we need this?

File: mlfinlab/tests/test_hrp.py [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+    def setUp(self):
+        """
+        Set the file path for the tick data csv
+        """
+        project_path = os.path.dirname(__file__)
+        data_path = project_path + '/test_data/stock_prices.csv'
+        self.data = pd.read_csv(data_path, parse_dates = True, index_col = "date")
+
+    def test_hrp(self):
+        hrp = HierarchicalRiskParity()
```
Comment (Jackal08) [inline mlfinlab/tests/test_hrp.py]: Please add docstrings....

File: mlfinlab.egg-info/SOURCES.txt [hunk 23]
```diff
@@ -0,0 +1,116 @@
+.coveragerc
```
Comment (Jackal08) [inline mlfinlab.egg-info/SOURCES.txt]: Please dont commit this file. Please delete.

File: mlfinlab.egg-info/dependency_links.txt [hunk 24]
```diff
@@ -0,0 +1 @@
+
```
Comment (Jackal08) [inline mlfinlab.egg-info/dependency_links.txt]: delete

File: mlfinlab.egg-info/top_level.txt [hunk 25]
```diff
@@ -0,0 +1 @@
+mlfinlab
```
Comment (Jackal08) [inline mlfinlab.egg-info/top_level.txt]: del

File: mlfinlab/portfolio_optimization/cla.py [hunk 26]
```diff
@@ -0,0 +1,608 @@
...
+Portfolio Optimization”,Algorithms, 6 (2013), 169-196. http://dx.doi.org/10.3390/a6010169
+'''
+
+import numbers
+from math import log, ceil
+import numpy as np
+import pandas as pd
+
+
+def _infnone(number):
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/cla.py]: Why is this logic outside the class? Can we put it in the class, if it is the only place it is used?

Comment (aditya1702) [reply mlfinlab/portfolio_optimization/cla.py]: @Jackal08 I already put it inside the class but havent commited. Will be reflected in the next commit

File: mlfinlab/portfolio_optimization/cla.py [hunk 27]
```diff
@@ -0,0 +1,608 @@
...
+    def _reduce_matrix(matrix, row_indices, col_indices):
+        '''
+        Reduce a matrix to the provided set of rows and columns
+        '''
+
+        return matrix[np.ix_(row_indices, col_indices)]
+
+    def _purge_num_err(self, tol):
+        '''
+        Purge violations of inequality constraints (associated with ill-conditioned cov matrix)
```
Comment (Jackal08) [inline mlfinlab/portfolio_optimization/cla.py]: Docstrings... Please always add docs for arguments.

File: mlfinlab/tests/test_portfolio_optimisation.py [hunk 28]
```diff
@@ -0,0 +1,274 @@
...
+        """
+        Set the file path for the tick data csv
+        """
+        project_path = os.path.dirname(__file__)
+        data_path = project_path + '/test_data/stock_prices.csv'
+        self.data = pd.read_csv(data_path, parse_dates=True, index_col="Date")
+
+    def test_hrp(self):
+        """
+        Test the HRP algorithm
```
Comment (Jackal08) [inline mlfinlab/tests/test_portfolio_optimisation.py]: Sure but how are you testing HRP?

File: mlfinlab/portfolio_optimization/cla.py (hunk 27 above)
Comment (aditya1702) [reply mlfinlab/portfolio_optimization/cla.py]: Resolved

File: mlfinlab/tests/test_portfolio_optimisation.py (hunk 28 above)
Comment (aditya1702) [reply mlfinlab/tests/test_portfolio_optimisation.py]: Resolved

**Pull request merged**